TombEngine/Objects/TR4/Trap/SquishyBlock.cpp

```diff
@@ -0,0 +1,285 @@
+	{
+		auto projectedPos = Geometry::TranslatePoint(item.Pose.Position, dir, BLOCK(0.5f));
+		auto pointColl = GetCollision(item.Pose.Position, item.RoomNumber, dir, BLOCK(0.5f));
+		auto bounds = GameBoundingBox(&item);
+		int itemHeight = bounds.GetHeight();
+		
+		// Test for wall.
+		if (pointColl.Block->IsWall(projectedPos.x, projectedPos.z))
+			return false;
+
+		// Test for slippery slope.
+		if (pointColl.Position.FloorSlope)
+			return false;
+
+		// Flat floor.
+		if ((abs(pointColl.FloorTilt.x) == 0 && abs(pointColl.FloorTilt.y) == 0))
+		{
+			// Test for step.
+			int relFloorHeight = abs(pointColl.Position.Floor - item.Pose.Position.y);
+			if (relFloorHeight >= CLICK(1) && item.Pose.Position.y >= pointColl.Position.Floor)
+				return false;
+		}
+		// Sloped floor.
+		else
+		{
+			// Half block.
+			int relFloorHeight = abs(pointColl.Position.Floor - item.Pose.Position.y);
+			if (relFloorHeight > CLICK(1))
+				return false;
+
+			short slopeAngle = ANGLE(0.0f);
+			if (pointColl.FloorTilt.x > 0)
+			{
+				slopeAngle = ANGLE(-90.0f);
+			}
+			else if (pointColl.FloorTilt.x < 0)
+			{
+				slopeAngle = ANGLE(90.0f);
+			}
+
+			if (pointColl.FloorTilt.y > 0 && pointColl.FloorTilt.y > abs(pointColl.FloorTilt.x))
+			{
+				slopeAngle = ANGLE(180.0f);
+			}
+			else if (pointColl.FloorTilt.y < 0 && -pointColl.FloorTilt.y > abs(pointColl.FloorTilt.x))
+			{
+				slopeAngle = ANGLE(0.0f);
+			}
+
+			short dirAngle = phd_atan(dir.z, dir.x);
+			short alignAngle = Geometry::GetShortestAngle(slopeAngle, dirAngle);
+
+			// Test if slope aspect is aligned with direction.
+			if (alignAngle != 0 && alignAngle != ANGLE(180.0f))
+				return false;
+		}
+
+		// Check for diagonal split.
+		if (pointColl.Position.DiagonalStep)
+			return false;
+
+		// Test ceiling height.
+		int relCeilHeight = abs(pointColl.Position.Ceiling - pointColl.Position.Floor);
+		
+		if (relCeilHeight <= itemHeight)
+			return false;
+
+		// Check for blocked grey box.
+		if (g_Level.Boxes[pointColl.Block->Box].flags & BLOCKABLE)
+		{
+			if (g_Level.Boxes[pointColl.Block->Box].flags & BLOCKED)
+				return false;
+		}
+
+		// Check for inaccessible sector.
+		if (pointColl.Block->Box == NO_ZONE)
+			return false;
+
+		// Check for stopper flag.
+		if (pointColl.Block->Stopper)
+			return false;
+
+		return true;
+	}
+
+	void SquishyBlockCollision(short itemNumber, ItemInfo* laraItem, CollisionInfo* coll)
+	{
+		auto& item = g_Level.Items[itemNumber];
+
+		if (TestBoundsCollide(&item, laraItem, coll->Setup.Radius) && TestCollision(&item, laraItem))
+		{
+			if (laraItem->HitPoints > 0)
+				ItemPushItem(&item, laraItem, coll, false, 1);			
+		}
+
+		if (ItemPushItem(&item, laraItem, coll, false, 1))
+		{
+			if (item.Animation.ActiveState == SQUISHY_BLOCK_STATE_ORIGINAL)
+			{
+				auto frame = item.Animation.FrameNumber - GetAnimData(item).frameBase;
+				if (!frame || frame == 33)
+					LaraItem->HitPoints = 0;
```

Comment:
Please use `DoDamage(&laraItem, INT_MAX)` here instead. Also, avoid referencing `LaraItem` global (note that `LaraItem` starts with a capital L here) in your code, it is "toxic" global we should eventually get rid of.